pymc3/step_methods/sgmcmc.py

```diff
@@ -0,0 +1,295 @@
+from collections import OrderedDict
+import warnings
+
+from .arraystep import Competence, ArrayStepShared
+from ..vartypes import continuous_types
+from ..model import modelcontext, inputvars
+import theano.tensor as tt
+from ..theanof import tt_rng, make_shared_replacements
+import theano
+import numpy as np
+
+__all__ = ['SGFS']
+
+EXPERIMENTAL_WARNING = "Warning: Stocastic Gradient based sampling methods are experimental step methods and not yet"\
+    " recommended for use in PyMC3!"
+
+def _value_error(cond, str):
+    """Throws ValueError if cond is False"""
+    if not cond:
+        raise ValueError(str)
+
+def _check_minibatches(minibatch_tensors, minibatches):
+    _value_error(isinstance(minibatch_tensors, list),
+                 'minibatch_tensors must be a list.')
+
+    _value_error(hasattr(minibatches, "__iter__"),
+                 'minibatches must be an iterator.')
+
+def prior_dlogp(vars, model, flat_view):
+    """Returns the gradient of the prior on the parameters as a vector of size D x 1"""
+    terms = tt.concatenate(
+        [theano.grad(var.logpt, var).flatten() for var in vars],
+        axis=0)
+    dlogp = theano.clone(terms, flat_view.replacements, strict=False)
+
+    return dlogp
+
+def elemwise_dlogL(vars, model, flat_view):
+    """
+    Returns Jacobian of the log likelihood for each training datum wrt vars
+    as a matrix of size N x D
+    """
+    # select one observed random variable 
+    obs_var = model.observed_RVs[0]
+    # tensor of shape (batch_size,)
+    logL = obs_var.logp_elemwiset.sum(axis=tuple(range(1, obs_var.logp_elemwiset.ndim)))
+    # calculate fisher information
+    terms = []
+    for var in vars:
+         output, _ =  theano.scan(lambda i, logX=logL, v=var: theano.grad(logX[i], v).flatten(),\
+                            sequences=[tt.arange(logL.shape[0])])
+         terms.append(output)
+    dlogL = theano.clone(tt.concatenate(terms, axis=1), flat_view.replacements, strict=False)
+    return dlogL
+
+
+class BaseStochasticGradient(ArrayStepShared):
+    R"""
+    BaseStochasticGradient Object
+    
+    For working with BaseStochasticGradient Object
+    we need to supply the probabilistic model 
+    (:code:`model`) with the data supplied to observed
+    variables of type `GeneratorOp`
+    
+    Parameters
+    -------
+    vars : list
+        List of variables for sampler
+    batch_size`: int
+        Batch Size for each step
+    total_size : int
+        Total size of the training data
+    step_size : float
+        Step size for the parameter update
+    model : PyMC Model
+        Optional model for sampling step. Defaults to None (taken from context)
+    random_seed : int
+        The seed to initialize the Random Stream
+    minibatches : iterator
```

Comment:
There are two issues I am running into which is preventing me from using pm.Minibatch

1. The batch size parameter is not configurable. I have different sized datasets for training and testing. I cannot use the same variable and change the data for running ppc after training.

2. The sampling speed slows down by 50% when using pm.Minibatch over using a simple iterator.

Here I have included a screenshot. Let me know if there are any easy fixes that I am missing, otherwise I feel I can post this problem in a issue separately.

<img width="577" alt="screen shot 2017-07-17 at 9 54 04 am" src="https://user-images.githubusercontent.com/5184832/28279705-2a3adc76-6ad6-11e7-8236-46d5add567d4.png">